Make sure all cells on LCAD card have an IMP:<particle> card if any of them do

ljacobson64 opened the pull request:
Trying this again

makeclean commented:
@ljacobson64 could you rebase against develop and then push again?

ljacobson64 commented:
Done

ljacobson64 commented:
Wow, this new Docker CI is really fast

makeclean commented:
That's the idea :)

gonuke commented:
Do we know why this fails in CI?

ljacobson64 commented:
Based on the fact that the only check that failed was the one with the master branch of MOAB, I'm assuming it's because I hadn't rebased this since the singleton was removed.

gonuke commented on mcnp/mcnp_funcs.cpp:
```diff
@@ -476,16 +476,38 @@ void write_lcad_old(std::ofstream &lcadfile)
+  bool imp_n_needed = false, imp_p_needed = false, imp_e_needed = false;
+  for( int i = 1; i <= num_cells; ++i ) {
+    moab::EntityHandle vol = DAG->entity_by_index( 3, i );
+    if( DAG->has_prop( vol, "imp.n" )) {
```
Is there a way to search for the existence of this tag directly in MOAB without looping through the volumes?  (Maybe not)

makeclean replied:
I would like to change the way that metadata is handled by Dagmc in general, for example at loading store it all in some metadata class so we can query it more neartly. But right now, one has use has_prop.

gonuke commented on mcnp/mcnp_funcs.cpp:
```diff
@@ -498,37 +520,53 @@ void write_lcad_old(std::ofstream &lcadfile)
 
-    lcadfile << cellid << " ";
+    // If no importances specified, default to neutron mode
+    if ( ! imp_n_needed && ! imp_p_needed && ! imp_e_needed ) {
```
makes more sense to move this up to just after you search for the tags and set `imp_*_needed` rather than inside loop for each cell

gonuke commented on mcnp/mcnp_funcs.cpp:
```diff
@@ -476,16 +476,38 @@ void write_lcad_old(std::ofstream &lcadfile)
+    moab::EntityHandle vol = DAG->entity_by_index( 3, i );
+    if( DAG->has_prop( vol, "imp.n" )) {
+      imp_n_needed = true;
+      cimp_n = 1;
```
reuse of the `cimp_*` variables makes it difficult to understand their intent, and thus to review the method.  Here it looks like they serve as default values, but then later they switch to complement values?  After copying the defaults elsewhere?

gonuke commented on mcnp/mcnp_funcs.cpp:
```diff
@@ -498,37 +520,53 @@ void write_lcad_old(std::ofstream &lcadfile)
-        std::cout << "Detected material and density specified for implicit complement: " << cmat <<", " << crho << std::endl;
-        cimp = imp_n;
+        std::cout << "Detected material and density specified for implicit complement: " << cmat << ", " << crho << std::endl;
+        if( imp_n_needed ) cimp_n = imp_n;
```
Doesn't this change the "default" values that are reused in line 508 in another iteration of the loop?

ljacobson64 commented:
There is a newer version of `mcnp_funcs.cpp` in my MCNP6 branch [here](https://github.com/ljacobson64/DAGMC/blob/mcnp6/mcnp/mcnp_funcs.cpp#L469) that addresses some of these issues.

Instead of having separate variables for `n`, `p`, and `e`, this version contains maps so that all the particle types can be handled. The `cimp` map has two purposes here: one is to specify the importances that should be assigned to the implicit complement, and the other is to keep track of which particles exist anywhere in the problem.

Perhaps it would be best to close this PR and open one for the MCNP6 branch.
